adding my own shitty gtfs csv parser + schedule files

jakswa opened the pull request:
Well, it's been a year and marta.io still suffers from the same API edge cases for terminating stations (and those stations near them). Here we have a calendar file that lays out the service # to use each day of the week, and then a file for each service. This schedule data is intended for use whenever a train isn't in the API for a given direction. I hope to highlight schedule data vs realtime data, so people know that it's not based on realtime data.

There should also exist exceptions for holidays/etc, but none exist in the GTFS spec right now, and I'm going to add those (and the logic for them) when they appear.

An alternative I considered was scraping MARTA's website for each station's schedule. I chose GTFS because it's a standard spec, and because I got scared while looking at the source code on MARTA's website.

- parser reduces 50MB or so of [GTFS CSVs](https://developers.google.com/transit/gtfs/reference) down to around 400kB worth of train-related json files
- this was as painful as I thought it'd be
- i'll need to re-run (and probably fix) this task as the gtfs data changes

bblack commented on lib/tasks/parse_gtfs.js:
```diff
@@ -0,0 +1,154 @@
+
+// of course they couldn't use the same station names
+// as in the realtime API
+var station_name_mapping = {
```
:smirk:

bblack commented on lib/tasks/parse_gtfs.js:
```diff
@@ -0,0 +1,154 @@
+    resolve = res;
+  });
+  var stop_ids = {};
+  // this file is 30-something-goddamn megabytes, guess I'll stream this one
```
highly recommend https://github.com/wdavidw/node-csv-parse - it's the parsing piece of `node-csv` and it's designed to read streams.

jakswa replied:
Nice, I was trying to avoid parsing all 30MB into memory. This looks like what I want. I'll convert to use it, seeing as I've also been relying on bad things like no quotes existing in the CSVs. :tropical_drink:

jakswa commented:
In my testing over the past couple of days, it turns out the schedule data is kind of useless in predicting when a train will depart from one of the terminal stations. It might still be better than nothing, but I'm leaning towards only using it on terminal stations like Doraville. 

Maybe 1 out of every 4 or 5 predictions is accurate enough, in my guestimation, to be useful. When things go well, you see these gray numbers that indicate schedule data. The as the countdown nears zero for doraville or airport, you'd see a train suddenly appear:

![m3](https://cloud.githubusercontent.com/assets/137793/6398817/45a83950-bdbc-11e4-921c-a57688ac1cb0.png)

More often than not, though, either the schedule time will pass with no train leaving (and we just roll over to predict the next scheduled arrival, which is depressing), or a train departs *too early* and we get confusing states going on. Below, the schedule still had 6min and 3min to go for the terminal stations, but a train left early, so we're stuck in some kind of useless state that I'm sure will confuse users:

![mart1](https://cloud.githubusercontent.com/assets/137793/6398747/a91e47e6-bdbb-11e4-87c7-2098191cad58.png)
![marta2](https://cloud.githubusercontent.com/assets/137793/6398748/a923bcee-bdbb-11e4-9398-3c9ecc377561.png)

bblack commented:
@itsmarta definitely has the realtime data for those end stations *somewhere*, since it's visible on the boards (at least at Chamblee). Have you tried contacting them to see if they can get that stuff reflected in the API?

http://www.itsmarta.com/developers/contact-us.aspx

I see you started this thread - you may want to push harder on it:
https://groups.google.com/forum/?fromgroups=#!topic/atl-transit-developers/JB0YWwgTHdM

Also the page at http://itsmarta.com/developers/default.aspx gives the email address martadevreq@itsmarta.com

jakswa commented:
Back when I was gathering feedback [from reddit](http://www.reddit.com/r/Atlanta/comments/20w32v/need_more_feature_requests_for_martaio/cg7l275), someone did PM me with an undocumented API route that is station specific (think `next_train/:station`) that outputs realtime data mixed with schedule data. The guess was that they used this endpoint for all the signs at each station, because it contains arrivals that are marked with `(scheduled)`. [Here's some example output for doraville.](https://gist.github.com/jakswa/2ede3e518362bcc16552)

However, I'd have to hit that endpoint once for each station we care about, and then repeated every 10s or so, if someone is looking at the website. Not ideal... but I could start doing it every 10s for the terminating stations, at least...

jakswa commented:
I'll deploy this branch to marta.io when I get a free moment later today.

jakswa commented:
This has been deployed for a couple days now and is working pretty well. I've cleaned up some things, fixed some edge cases, added `csv-parse`, and updated the README. Clickin' the button!

![img](http://replygif.net/i/773.gif)
